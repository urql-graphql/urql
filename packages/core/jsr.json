{
  "name": "@urql/core",
  "version": "5.0.6",
  "exports": {
    ".": "./src/index.ts",
    "./internal": "./src/internal/index.ts"
  },
  "exclude": [
    "node_modules",
    "cypress",
    "**/*.test.*",
    "**/*.spec.*",
    "**/*.test.*.snap",
    "**/*.spec.*.snap"
  ]
}
