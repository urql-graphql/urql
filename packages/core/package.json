{
  "name": "@urql/core",
  "version": "3.0.3",
  "description": "The shared core for the highly customizable and versatile GraphQL client",
  "sideEffects": false,
  "homepage": "https://formidable.com/open-source/urql/docs/",
  "bugs": "https://github.com/FormidableLabs/urql/issues",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/FormidableLabs/urql.git",
    "directory": "packages/core"
  },
  "keywords": [
    "graphql client",
    "state management",
    "cache",
    "formidablelabs",
    "exchanges"
  ],
  "main": "dist/urql-core",
  "module": "dist/urql-core.mjs",
  "types": "dist/types/index.d.ts",
  "source": "src/index.ts",
  "exports": {
    ".": {
      "import": "./dist/urql-core.mjs",
      "require": "./dist/urql-core.js",
      "types": "./dist/types/index.d.ts",
      "source": "./src/index.ts"
    },
    "./package.json": "./package.json",
    "./internal": {
      "import": "./dist/urql-core-internal.mjs",
      "require": "./dist/urql-core-internal.js",
      "types": "./dist/types/internal/index.d.ts",
      "source": "./src/internal/index.ts"
    }
  },
  "files": [
    "LICENSE",
    "README.md",
    "dist/",
    "internal/"
  ],
  "scripts": {
    "test": "jest",
    "clean": "rimraf dist",
    "check": "tsc --noEmit",
    "lint": "eslint --ext=js,jsx,ts,tsx .",
    "build": "rollup -c ../../scripts/rollup/config.js",
    "prepare": "node ../../scripts/prepare/index.js",
    "prepublishOnly": "run-s clean build"
  },
  "jest": {
    "preset": "../../scripts/jest/preset"
  },
  "devDependencies": {
    "graphql": "^16.0.0"
  },
  "peerDependencies": {
    "graphql": "^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0 || ^15.0.0 || ^16.0.0"
  },
  "dependencies": {
    "@graphql-typed-document-node/core": "^3.1.1",
    "wonka": "^6.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
