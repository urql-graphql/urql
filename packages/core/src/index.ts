export { gql } from './gql';

export * from './client';
export * from './exchanges';
export * from './types';

export {
  CombinedError,
  stringifyVariables,
  createRequest,
  makeResult,
  makeErrorResult,
  mergeResultPatch,
  formatDocument,
  maskTypename,
  makeOperation,
  getOperationName,
} from './utils';
