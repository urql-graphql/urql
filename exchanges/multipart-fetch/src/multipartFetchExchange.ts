import { filter, merge, mergeMap, pipe, share, takeUntil, onPush } from 'wonka';
import { extractFiles } from 'extract-files';
import { Exchange } from '@urql/core';

import {
  makeFetchBody,
  makeFetchURL,
  makeFetchOptions,
  makeFetchSource,
} from '@urql/core/internal';

export const multipartFetchExchange: Exchange = ({
  forward,
  dispatchDebug,
}) => ops$ => {
  const sharedOps$ = share(ops$);
  const fetchResults$ = pipe(
    sharedOps$,
    filter(operation => {
      return operation.kind === 'query' || operation.kind === 'mutation';
    }),
    mergeMap(operation => {
      const teardown$ = pipe(
        sharedOps$,
        filter(op => op.kind === 'teardown' && op.key === operation.key)
      );

      // Spreading operation.variables here in case someone made a variables with Object.create(null).
      const { files, clone: variables } = extractFiles({
        ...operation.variables,
      });
      const body = makeFetchBody({ query: operation.query, variables });

      let url: string;
      let fetchOptions: RequestInit;
      if (files.size) {
        url = makeFetchURL(operation);
        fetchOptions = makeFetchOptions(operation);
        if (fetchOptions.headers!['content-type'] === 'application/json') {
          delete fetchOptions.headers!['content-type'];
        }

        fetchOptions.method = 'POST';
        fetchOptions.body = new FormData();
        fetchOptions.body.append('operations', JSON.stringify(body));

        const map = {};
        let i = 0;
        files.forEach(paths => {
          map[++i] = paths.map(path => `variables.${path}`);
        });

        fetchOptions.body.append('map', JSON.stringify(map));

        i = 0;
        files.forEach((_, file) => {
          (fetchOptions.body as FormData).append(`${++i}`, file, file.name);
        });
      } else {
        url = makeFetchURL(operation, body);
        fetchOptions = makeFetchOptions(operation, body);
      }

      dispatchDebug({
        type: 'fetchRequest',
        message: 'A fetch request is being executed.',
        operation,
        data: {
          url,
          fetchOptions,
        },
      });

      return pipe(
        makeFetchSource(operation, url, fetchOptions),
        takeUntil(teardown$),
        onPush(result => {
          const error = !result.data ? result.error : undefined;

          dispatchDebug({
            type: error ? 'fetchError' : 'fetchSuccess',
            message: `A ${
              error ? 'failed' : 'successful'
            } fetch response has been returned.`,
            operation,
            data: {
              url,
              fetchOptions,
              value: error || result,
            },
          });
        })
      );
    })
  );

  const forward$ = pipe(
    sharedOps$,
    filter(operation => {
      return operation.kind !== 'query' && operation.kind !== 'mutation';
    }),
    forward
  );

  return merge([fetchResults$, forward$]);
};
