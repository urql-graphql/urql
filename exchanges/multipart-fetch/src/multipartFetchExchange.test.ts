import { Client, OperationResult, makeOperation } from '@urql/core';
import { empty, fromValue, pipe, Source, subscribe, toPromise } from 'wonka';

import { multipartFetchExchange } from './multipartFetchExchange';

import {
  uploadOperation,
  queryOperation,
  multipleUploadOperation,
} from './test-utils';

const fetch = (global as any).fetch as jest.Mock;
const abort = jest.fn();

const abortError = new Error();
abortError.name = 'AbortError';

beforeAll(() => {
  (global as any).AbortController = function AbortController() {
    this.signal = undefined;
    this.abort = abort;
  };
});

afterEach(() => {
  fetch.mockClear();
  abort.mockClear();
});

afterAll(() => {
  (global as any).AbortController = undefined;
});

const response = JSON.stringify({
  status: 200,
  data: {
    data: {
      user: 1200,
    },
  },
});

const exchangeArgs = {
  forward: () => empty as Source<OperationResult>,
  client: {} as Client,
  dispatchDebug: jest.fn(),
};

describe('on success', () => {
  beforeEach(() => {
    fetch.mockResolvedValue({
      status: 200,
      text: jest.fn().mockResolvedValue(response),
    });
  });

  it('uses a file when given', async () => {
    const fetchOptions = jest.fn().mockReturnValue({});

    const data = await pipe(
      fromValue({
        ...uploadOperation,
        context: {
          ...uploadOperation.context,
          fetchOptions,
        },
      }),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data).toMatchSnapshot();
    expect(fetchOptions).toHaveBeenCalled();
    expect(fetch.mock.calls[0][1].headers).toMatchSnapshot();
    expect(fetch.mock.calls[0][1].body).toMatchSnapshot();
  });

  it('uses multiple files when given', async () => {
    const fetchOptions = jest.fn().mockReturnValue({});

    const data = await pipe(
      fromValue({
        ...multipleUploadOperation,
        context: {
          ...multipleUploadOperation.context,
          fetchOptions,
        },
      }),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data).toMatchSnapshot();
    expect(fetchOptions).toHaveBeenCalled();
    expect(fetch.mock.calls[0][1].headers).toMatchSnapshot();
    expect(fetch.mock.calls[0][1].body).toMatchSnapshot();
  });

  it('returns response data', async () => {
    const fetchOptions = jest.fn().mockReturnValue({});

    const data = await pipe(
      fromValue({
        ...queryOperation,
        context: {
          ...queryOperation.context,
          fetchOptions,
        },
      }),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data).toMatchSnapshot();
    expect(fetchOptions).toHaveBeenCalled();
    expect(fetch.mock.calls[0][1].body).toMatchSnapshot();
  });
});

describe('on error', () => {
  beforeEach(() => {
    fetch.mockResolvedValue({
      status: 400,
      text: jest.fn().mockResolvedValue('{}'),
    });
  });

  it('returns error data', async () => {
    const data = await pipe(
      fromValue(queryOperation),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data).toMatchSnapshot();
  });

  it('returns error data with status 400 and manual redirect mode', async () => {
    const fetchOptions = jest.fn().mockReturnValue({ redirect: 'manual' });

    const data = await pipe(
      fromValue({
        ...queryOperation,
        context: {
          ...queryOperation.context,
          fetchOptions,
        },
      }),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data).toMatchSnapshot();
  });

  it('ignores the error when a result is available', async () => {
    fetch.mockResolvedValue({
      status: 400,
      text: jest.fn().mockResolvedValue(response),
    });

    const data = await pipe(
      fromValue(queryOperation),
      multipartFetchExchange(exchangeArgs),
      toPromise
    );

    expect(data.data).toEqual(JSON.parse(response).data);
  });
});

describe('on teardown', () => {
  it('does not start the outgoing request on immediate teardowns', () => {
    fetch.mockRejectedValueOnce(abortError);

    const { unsubscribe } = pipe(
      fromValue(queryOperation),
      multipartFetchExchange(exchangeArgs),
      subscribe(fail)
    );

    unsubscribe();
    expect(fetch).toHaveBeenCalledTimes(0);
    expect(abort).toHaveBeenCalledTimes(1);
  });

  it('aborts the outgoing request', async () => {
    fetch.mockRejectedValueOnce(abortError);

    const { unsubscribe } = pipe(
      fromValue(queryOperation),
      multipartFetchExchange(exchangeArgs),
      subscribe(fail)
    );

    await Promise.resolve();

    unsubscribe();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(abort).toHaveBeenCalledTimes(1);
  });

  it('does not call the query', () => {
    pipe(
      fromValue(
        makeOperation('teardown', queryOperation, queryOperation.context)
      ),
      multipartFetchExchange(exchangeArgs),
      subscribe(fail)
    );

    expect(fetch).toHaveBeenCalledTimes(0);
    expect(abort).toHaveBeenCalledTimes(0);
  });
});
