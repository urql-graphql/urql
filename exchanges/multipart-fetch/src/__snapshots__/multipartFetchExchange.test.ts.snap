// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on error returns error data 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on error returns error data with status 400 and manual redirect mode 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "redirect": "manual",
            },
          },
        ],
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on success returns response data 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {},
          },
        ],
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on success returns response data 2`] = `"{\\"query\\":\\"query getUser($name: String) {\\\\n  user(name: $name) {\\\\n    id\\\\n    firstName\\\\n    lastName\\\\n  }\\\\n}\\\\n\\",\\"operationName\\":\\"getUser\\",\\"variables\\":{\\"name\\":\\"Clara\\"}}"`;

exports[`on success uses a file when given 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {},
          },
        ],
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 3,
    "kind": "mutation",
    "query": Object {
      "__key": 4034972436,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "uploadProfilePicture",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "picture",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "picture",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "uploadProfilePicture",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "location",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "File",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "picture",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 125,
        "source": Object {
          "body": "# uploadProfilePicture
mutation uploadProfilePicture($picture: File) { uploadProfilePicture(picture: $picture) { location } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "picture": File {},
    },
  },
}
`;

exports[`on success uses a file when given 2`] = `Object {}`;

exports[`on success uses a file when given 3`] = `FormData {}`;

exports[`on success uses multiple files when given 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {},
          },
        ],
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 3,
    "kind": "mutation",
    "query": Object {
      "__key": 2033658603,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "uploadProfilePictures",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "pictures",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "pictures",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "uploadProfilePicture",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "location",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "ListType",
                "loc": undefined,
                "type": Object {
                  "kind": "NamedType",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "File",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "pictures",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 132,
        "source": Object {
          "body": "# uploadProfilePictures
mutation uploadProfilePictures($pictures: [File]) { uploadProfilePicture(pictures: $pictures) { location } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "picture": Array [
        File {},
        File {},
      ],
    },
  },
}
`;

exports[`on success uses multiple files when given 2`] = `Object {}`;

exports[`on success uses multiple files when given 3`] = `FormData {}`;
